# Load Data from Google Drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Data

In [ ]:
import pandas as pd
import re
from datetime import datetime

# load data from our output result table
df = pd.read_excel("/content/drive/MyDrive/UPM/AIHeathcare2023/Data/D1.xlsx", header = 0)

In [ ]:
df.head(5)

PN  \
0                    US2021350883-A1   
1      CN113871020-A;  CN113871020-B   
2                   IN202111045992-A   
3                   IN202141008949-A   
4  US2021375462-A1;  WO2021247733-A1   

                                                  TI  \
0  User and test sample authentication method for...   
1  Method for managing health of emergency critic...   
2  Intelligent healthcare system useful for ident...   
3  Artificial intelligence (AI) enabled framework...   
4  Providing remote medical assessment of individ...   

                                                  AU  \
0                                        WU X;  LI S   
1                                LI Y;  LIU W;  LU L   
2  BATHLA R K;  ARORA P;  ARORA N;  KUMAR V;  PAN...   
3             MUDDAPU P;  NEELAKANTAPPA M;  MANYAM R   
4                            DORAZIO M A;  RAYOS J C   

                                                  AE          GA  \
0  WU X (WUXX-Individual);  LI S (LISS-Individual...  2021C8573T   
1  BEIJING ZHUOJIAN TECHNOLOGY CO LTD (BEIJ-Non-s...  202217124U   
2  BATHLA R K (BATH-Individual);  ARORA P (AROR-I...  2021D4631E   
3  MUDDAPU P (MUDD-Individual);  NEELAKANTAPPA M ...  202129332L   
4  DORAZIO M A (DORA-Individual);  RAYOS J C (RAY...  2021D9078A   

                                                  AB  TF   EA  \
0  NOVELTY - User and test sample authentication ... NaN  NaN   
1  NOVELTY - The method involves constructing a h... NaN  NaN   
2  NOVELTY - Intelligent healthcare system compri... NaN  NaN   
3  NOVELTY - The framework exploits both supervis... NaN  NaN   
4  NOVELTY - Method (M1) for providing remote med... NaN  NaN   

                                                  DC  \
0  D16 (Fermentation industry - including ferment...   
1  S05 (Electrical Medical Equipment);  T01 (Digi...   
2  B04 (Natural products and polymers. Including ...   
3                            T01 (Digital Computers)   
4  P31 (Diagnosis, surgery (A61B).);  S05 (Electr...   

                                                  MC  ...  \
0  D05-H09;  B11-C14;  B12-K04G1B;  B11-C08J;  B0...  ...   
1  S05-G02G9;  S05-G02G4;  S05-G02G1;  S05-D01;  ...  ...   
2  B04-F02A;  B11-C08J;  B11-C08L;  B11-C11;  B12...  ...   
3  T01-J05B2;  T01-J05B3;  T01-J10B2;  T01-J16B; ...  ...   
4  P31-A05;  S05-Y01;  S05-G02G2;  S05-D06;  S05-...  ...   

                                                  PI  \
0                            US868031    06 May 2020   
1                          CN11128954    26 Sep 2021   
2                          IN11045992    08 Oct 2021   
3                          IN41008949    03 Mar 2021   
4  US033698P    02 Jun 2020;  US337142    02 Jun ...   

                                                  DS   FS  \
0                                                NaN  NaN   
1                                                NaN  NaN   
2                                                NaN  NaN   
3                                                NaN  NaN   
4  WO2021247733-A1:      (National): AE;  AG;  AL...  NaN   

                                                  CP  \
0  US2021350883-A1 -- US20200159957-A1   ;  US202...   
1  CN113871020-A -- AU9953551-A   SMITHKLINE BEEC...   
2                                                NaN   
3                                                NaN   
4  ;  WO2021247733-A1 -- US20080170777-A1   ;  US...   

                                                  CR   DN   MN   RI   CI   RG  
0                                                NaN  NaN  NaN  NaN  NaN  NaN  
1                                                NaN  NaN  NaN  NaN  NaN  NaN  
2                                                NaN  NaN  NaN  NaN  NaN  NaN  
3                                                NaN  NaN  NaN  NaN  NaN  NaN  
4  WO2021247733-A1  LI LIN, QIN LIXIN, XU ZEGUO, ...  NaN  NaN  NaN  NaN  NaN  

[5 rows x 24 columns]

In [ ]:
df.shape

(1391, 24)

In [ ]:
# Filter rows based on the condition of US-patent
df = df[df['PN'].str.contains('US')]

df.head(5)

PN  \
0                                     US2021350883-A1   
4                   US2021375462-A1;  WO2021247733-A1   
5   WO2019177451-A1;  AU2019236012-A1;  KR20201218...   
7                                     US2021365815-A1   
12                    US2017169177-A1;  US10325070-B2   

                                                   TI  \
0   User and test sample authentication method for...   
4   Providing remote medical assessment of individ...   
5   System for facilitating diagnosis and treatmen...   
7   System for optimizing/personalizing user      ...   
12  Integrated computer-implemented system for    ...   

                                        AU  \
0                              WU X;  LI S   
4                  DORAZIO M A;  RAYOS J C   
5         CHEE CHONG T;  CHEE C T;  CHEE C   
7   BIERMAN T M;  BEYERS J E;  BONUTTI P M   
12                 BEALE R G;  JARUZEL M E   

                                                   AE          GA  \
0   WU X (WUXX-Individual);  LI S (LISS-Individual...  2021C8573T   
4   DORAZIO M A (DORA-Individual);  RAYOS J C (RAY...  2021D9078A   
5   T-BIOMAX SDN BHD (TBIO-Non-standard);  T-BIOMA...  2019811792   
7                      P TECH LLC (PTEP-Non-standard)  2021D48667   
12  BEALE R G (BEAL-Individual);  JARUZEL M E (JAR...  2017405002   

                                                   AB  TF   EA  \
0   NOVELTY - User and test sample authentication ... NaN  NaN   
4   NOVELTY - Method (M1) for providing remote med... NaN  NaN   
5   NOVELTY - The system (100) has a database (120... NaN  NaN   
7   NOVELTY - The system has at least one sensor c... NaN  NaN   
12  NOVELTY - The system (100) has a computer-impl... NaN  NaN   

                                                   DC  \
0   D16 (Fermentation industry - including ferment...   
4   P31 (Diagnosis, surgery (A61B).);  S05 (Electr...   
5   P31 (Diagnosis, surgery (A61B).);  S05 (Electr...   
7   P31 (Diagnosis, surgery (A61B).);  S05 (Electr...   
12  T01 (Digital Computers);  W01 (Telephone and D...   

                                                   MC  ...  \
0   D05-H09;  B11-C14;  B12-K04G1B;  B11-C08J;  B0...  ...   
4   P31-A05;  S05-Y01;  S05-G02G2;  S05-D06;  S05-...  ...   
5   P31-A05;  S05-D01;  T01-J05B4B;  T01-J05B4F;  ...  ...   
7   P31-A05;  S05-G02G4;  S05-A01A1;  T01-J40;  T0...  ...   
12  T01-D01;  T01-J06A;  T01-J12C;  T01-J30A;  T01...  ...   

                                                   PI  \
0                             US868031    06 May 2020   
4   US033698P    02 Jun 2020;  US337142    02 Jun ...   
5   MY700974    12 Mar 2018;  MY701003    13 Mar 2...   
7   US552091P    30 Aug 2017;  US395177    05 Aug ...   
12  US266949P    14 Dec 2015;  US377490    13 Dec ...   

                                                   DS   FS  \
0                                                 NaN  NaN   
4   WO2021247733-A1:      (National): AE;  AG;  AL...  NaN   
5   WO2019177451-A1:      (National): AE;  AG;  AL...  NaN   
7                                                 NaN  NaN   
12                                                NaN  NaN   

                                                   CP  \
0   US2021350883-A1 -- US20200159957-A1   ;  US202...   
4   ;  WO2021247733-A1 -- US20080170777-A1   ;  US...   
5   WO2019177451-A1 -- US20010056228-A1   ;  US200...   
7   US2021365815-A1 -- US5589818-A   QUEEN A (QUEE...   
12  US2017169177-A1 -- US20050038669-A1   ;  US200...   

                                                   CR   DN   MN   RI   CI   RG  
0                                                 NaN  NaN  NaN  NaN  NaN  NaN  
4   WO2021247733-A1  LI LIN, QIN LIXIN, XU ZEGUO, ...  NaN  NaN  NaN  NaN  NaN  
5                                                 NaN  NaN  NaN  NaN  NaN  NaN  
7                                                 NaN  NaN  NaN  NaN  NaN  NaN  
12                                                NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 24 c

In [ ]:
df.shape

(1020, 24)

## US-patent Data

*   (Worldwide) 1391 patents
*   (US) 1020 patents

Dive deeper:

*   (Single) 405 patents
*   (Multiple) 615 patents



In [ ]:
# # Filter rows based on the condition
# df = df[~df['PN'].str.contains(';')]

## Data Cleaning

In [ ]:
df1 = df

In [ ]:
# Extract the first date from the 'PI' column and convert it to datetime
df1['Time'] = df1['PI'].str.extract('(\d{2}\s\w+\s\d{4})')
df1['Time'] = pd.to_datetime(df1['Time'], errors='coerce')

# # Extract the last date from the 'PI' column and convert it to datetime
# # Function to extract the last date from the 'PI' column
# def extract_last_date(row):
#     dates = [date.split()[-3:] for date in row.split(';')]
#     dates = [date for date in dates if len(date) == 3]
#     if dates:
#         return max([datetime.strptime(' '.join(date), '%d %b %Y') for date in dates])
#     else:
#         return pd.NaT

# # Create the 'Time' column by applying the function
# df1['Time'] = df1['PI'].apply(extract_last_date)


# Remove prefix "NOVELTY -" from the "AB" column
df1['AB'] = df1['AB'].str.replace('NOVELTY - ', '')

# Split the "AU" and "DC" columns based on the ";" delimiter
df1['AU'] = df1['AU'].str.split(';')
df1['DC'] = df1['DC'].str.split(';')

# Strip whitespaces from each string in the "AU" and "DC" columns
df1['AU'] = df1['AU'].apply(lambda x: [item.strip() for item in x] if isinstance(x, list) else [])
df1['DC'] = df1['DC'].apply(lambda x: [item.strip() for item in x] if isinstance(x, list) else [])


# 'TI' and 'AB' are the column we want to clean
df1['TI'] = df1['TI'].apply(lambda x: re.sub(r'\s+', ' ', str(x)) if isinstance(x, (str, bytes)) else x)
df1['AB'] = df1['AB'].apply(lambda x: re.sub(r'\s+', ' ', str(x)) if isinstance(x, (str, bytes)) else x)

# Assuming df1 is your DataFrame
df1 = df1.dropna(subset=['AB'])

# Assuming df1 is your DataFrame
df1 = df1.sort_values(by='Time', ascending=True)

# Drop rows with duplicate values in the "TI" column
df1 = df1.drop_duplicates(subset='TI', keep='first')

# Keep only the required columns
df1_cleaned = df1[['TI', 'AU', 'AB', 'DC', 'Time']]

# Filter the DataFrame to include only rows with dates within the specified range
start_date = '2011-01-01'
end_date = '2020-12-31'

df1_cleaned = df1_cleaned[df1_cleaned['Time'].between(start_date, end_date)]

# Display the resulting DataFrame
df1_cleaned.head(5)

TI  \
1302  Apparatus for detecting asymmetry in brain of ...   
1146  Wireless sensing terminal for use during treat...   
1319  Computer based system e.g. end-user computers,...   
591   System for use by e.g. clinician, to diagnose ...   
1130  Diagnosing colon cancer in sample obtained fro...   

                                                     AU  \
1302                             [THIELE F O, THIELE F]   
1146                                       [HASEGAWA Y]   
1319                                          [RAWAT R]   
591   [SWERDLOW C D, SACHANANDANI H G, GUNDERSON B D...   
1130  [HARKIN D P, PROUTSKI V, BLACK J, KERR P, KENN...   

                                                     AB  \
1302  The apparatus has a processor comparing corres...   
1146  The terminal (10) has a main power supply unit...   
1319  The system has a portion of a computer readabl...   
591   The system has an implantable medical device (...   
1130  A method (M1) for diagnosing colon cancer in a...   

                                                     DC       Time  
1302  [S03 (Scientific Instrumentation), S05 (Electr... 2011-01-05  
1146  [S05 (Electrical Medical Equipment), W02 (Broa... 2011-01-17  
1319  [S05 (Electrical Medical Equipment), T01 (Digi... 2011-01-18  
591   [P34 (Sterilising, syringes, electrotherapy (A... 2011-01-21  
1130  [B04 (Natural products and polymers. Including... 2011-01-25

In [ ]:
df1_cleaned.tail(5)

TI  \
654   Device for training battery usage models, has ...   
1313  Method for using health detection supportive d...   
547   Method for providing data catalog by computer ...   
183   System useful for monitoring airborne transmit...   
66    Device for providing health care service to e....   

                                                     AU  \
654                         [GUNTUKU K K, BOYAPATI V K]   
1313                               [HAREL D, SHARGAL S]   
547   [SHIN W T, AN H J, PARK S M, LEE J H, HWANG S ...   
183                                  [DCRUZ J G, YEH P]   
66                                   [SHIN M, SHIN M Y]   

                                                     AB  \
654   The device has a processor for generating or r...   
1313  The system has a sensor unit (1) with lateral ...   
547   The method involves generating first recommend...   
183   System comprises several sensors located in or...   
66    The device has a detachable unit which is conf...   

                                                     DC       Time  
654   [T01 (Digital Computers), W01 (Telephone and D... 2020-09-03  
1313  [P31 (Diagnosis, surgery (A61B).), S05 (Electr... 2020-10-16  
547                           [T01 (Digital Computers)] 2020-12-14  
183   [P31 (Diagnosis, surgery (A61B).), S05 (Electr... 2020-12-25  
66    [P31 (Diagnosis, surgery (A61B).), S05 (Electr... 2020-12-29

In [ ]:
df1_cleaned.shape

(629, 5)

## Data with time range:


*   629 patents from 2011 - 2020 (10 years)
*   (Optional): patents from 2021 - 2023 (3 years)


# Saving Data

In [ ]:
# Assuming df1_cleaned is your cleaned DataFrame
df1_cleaned.to_csv('/content/drive/MyDrive/UPM/SGE2024/SGE_D1_cleaned.csv', index=False)

# EDA report

In [ ]:
!pip install dataprep

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.9/9.9 MB 24.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.5/18.5 MB 27.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.6/133.6 kB 14.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 24.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 21.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 22.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 22.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 764.0/764.0 kB 27.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.4/6.4 MB 21.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.6/49.6 kB 7.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 7.3 MB/s eta 0:00:00
  Created wheel for sqlalchemy: filenam

In [ ]:
from dataprep.eda import create_report
from dataprep.eda import plot, plot_correlation, plot_missing

# create the cleaned EDA report
report = create_report(df1_cleaned, title='EDA Report on Patents')
report.save("/content/drive/MyDrive/UPM/SGE2024/patent_report")

Report has been saved to /content/drive/MyDrive/UPM/SGE2024/patent_report.html!


## Analysis

In [ ]:
df['length'] = df['AB'].apply(lambda x: len(str(x)))

min_length = df['length'].min()
max_length = df['length'].max()

print("Minimum length:", min_length)
print("Maximum length:", max_length)

Minimum length: 433
Maximum length: 7057


In [ ]:
df['word_count'] = df['AB'].apply(lambda x: len(str(x).split()))

# Get the minimum and maximum values of the number of words in the 'AB' column
min_word_count = df['word_count'].min()
max_word_count = df['word_count'].max()

print("Minimum number of words:", min_word_count)
print("Maximum number of words:", max_word_count)

Minimum number of words: 55
Maximum number of words: 1075


2 Modeling

---

